### PR TITLE
Create Dockerfile for PVPGN

### PR DESCRIPTION
**Pull request by aperture147**

### CODE REVIEW

**Comment by HarpyWar** [comment]
I think the Dockerfile must provide options to run a fully functional server. Otherwise it can be useful only for tests.
So it should have features, which can be described with the following examples:

1) Run pvpgn with plain storage (actually like in your current dockerfile) 
2) Run pvpgn with plain storage and mount `conf`, `files`, `lua`, `var`  to a host directory
3) Run pvpgn with mysql storage and pass parameters through environment variables: mysql_host, port, user, pass, db
4) Run pvpgn with mysql storage and connect it to another container with mysql server
5) `docker-compose.yml` which can run both pvpgn container + mysql container and mount directories to a host machine

Optionally it will be great to make a second docker image for d2cs + d2dbs, and third image for d2gs..

**Comment by HarpyWar** [comment]
https://github.com/pvpgn/pvpgn-server/issues/268

**Comment by aperture147** [comment]
I’m not familiar bash script though (mainly use python). Btw to maximize the customization, I prefer the mounting the configuration directory/configuration files method to env var. Currently I’m using the mounting method with SQL.
I will try to make an env var version using envsubstr. Btw is there anyway that I can build and pack the whole libraries so I can make thr multistage-build more easier?

**Comment by HarpyWar** [comment]
It should not be very hard to replace a line storage_path in bnetd.conf, it could be done with sed
```
sed -i 's/old-text/new-text/g'
```
All the different storage types for pvpgn are hardcoded in the executable file. Switch between them are done when run `cmake` with specified flags
https://github.com/pvpgn/pvpgn-server/blob/36f99253577487fdcf803183de29f70ff034c45e/CMakeLists.txt#L21-L24

**Comment by HarpyWar** [comment]
> Btw to maximize the customization, I prefer the mounting the configuration directory/configuration files method to env var. Currently I’m using the mounting method with SQL.

Could you explain with an example?

Generally mount is done through bind mounts `-v` or `--mount`. I'm not sure how it can be done better, for instance all the directories `var` `conf` `lua` `files` can be symlinked to a single place and then mounted.

**Comment by aperture147** [comment]
This is a small example I'm using in my server:

```yml
version: "3"

services:
    pvpgn:
        image: pvpgn
        command: bnetd -f
        restart: unless-stopped
        ports:
            - 6112:6112/udp
            - 6112:6112
            - 6200:6200/udp
            - 6200:6200
        volumes:
            - ./pvpgn/i18n:/usr/local/etc/pvpgn/i18n
            - ./pvpgn/bnetd.conf:/usr/local/etc/pvpgn/bnetd.conf
            - ./pvpgn/icons.conf:/usr/local/etc/pvpgn/icons.conf
            - ./pvpgn/channel.conf:/usr/local/etc/pvpgn/channel.conf
            - ./pvpgn/topics.conf:/usr/local/etc/pvpgn/topics.conf
            - ./pvpgn/tournament.conf:/usr/local/etc/pvpgn/tournament.conf
            - ./pvpgn/ad.json:/usr/local/etc/pvpgn/ad.json
            - ./pvpgn/files/mobavn01.png:/usr/local/var/pvpgn/files/mobavn01.png
            - ./pvpgn/files/mobavn02.png:/usr/local/var/pvpgn/files/mobavn02.png
            - ./pvpgn/ladders:/usr/local/var/pvpgn/ladders
```
and this is the tree of the the pvpgn directory.
```
pvpgn/
├── ad.json
├── bnetd.conf
├── channel.conf
├── files
│   ├── mobavn01.png
│   └── mobavn02.png
├── i18n
│   ├── bgBG
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── bnhelp.conf
│   ├── bnmotd.txt
│   ├── chathelp-war3.txt
│   ├── common.xml
│   ├── csCZ
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── deDE
│   │   ├── bnhelp.conf
│   │   ├── bnmotd.txt
│   │   ├── chathelp-war3.txt
│   │   ├── common.xml
│   │   ├── newaccount.txt
│   │   ├── termsofservice.txt
│   │   └── w3motd.txt
│   ├── esES
│   │   ├── bnhelp.conf
│   │   ├── bnmotd.txt
│   │   ├── chathelp-war3.txt
│   │   ├── common.xml
│   │   ├── newaccount.txt
│   │   ├── termsofservice.txt
│   │   └── w3motd.txt
│   ├── frFR
│   │   ├── bnmotd.txt
│   │   ├── chathelp-war3.txt
│   │   └── common.xml
│   ├── itIT
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── jpJA
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── koKR
│   │   ├── bnmotd.txt
│   │   ├── common.xml
│   │   └── termsofservice.txt
│   ├── newaccount.txt
│   ├── news.txt
│   ├── nlNL
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── plPL
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── ptBR
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── ruRU
│   │   ├── bnmotd.txt
│   │   ├── chathelp-war3.txt
│   │   └── common.xml
│   ├── svSE
│   │   ├── bnmotd.txt
│   │   └── common.xml
│   ├── termsofservice.txt
│   ├── w3motd.txt
│   ├── zhCN
│   │   ├── bnmotd.txt
│   │   ├── chathelp-war3.txt
│   │   └── common.xml
│   └── zhTW
│       ├── bnmotd.txt
│       └── common.xml
├── icons.conf
├── ladders
│   ├── RAL2_SOLO_LADDER
│   ├── SEXP_AG_LADDER
│   ├── SEXP_AR_LADDER
│   ├── SEXP_AW_LADDER
│   ├── SEXP_CG_LADDER
│   ├── SEXP_CR_LADDER
│   ├── SEXP_CW_LADDER
│   ├── STAR_AG_LADDER
│   ├── STAR_AR_LADDER
│   ├── STAR_AW_LADDER
│   ├── STAR_CG_LADDER
│   ├── STAR_CR_LADDER
│   ├── STAR_CW_LADDER
│   ├── TSUN_SOLO_LADDER
│   ├── TSXP_SOLO_LADDER
│   ├── W2BN_AGI_LADDER
│   ├── W2BN_AG_LADDER
│   ├── W2BN_ARI_LADDER
│   ├── W2BN_AR_LADDER
│   ├── W2BN_AWI_LADDER
│   ├── W2BN_AW_LADDER
│   ├── W2BN_CGI_LADDER
│   ├── W2BN_CG_LADDER
│   ├── W2BN_CRI_LADDER
│   ├── W2BN_CR_LADDER
│   ├── W2BN_CWI_LADDER
│   ├── W2BN_CW_LADDER
│   ├── W3XP_AT_LADDER
│   ├── W3XP_FFA_LADDER
│   ├── W3XP_SOLO_LADDER
│   ├── W3XP_TEAM_LADDER
│   ├── WAR3_AT_LADDER
│   ├── WAR3_FFA_LADDER
│   ├── WAR3_SOLO_LADDER
│   ├── WAR3_TEAM_LADDER
│   └── YURI_SOLO_LADDER
├── topics.conf
└── tournament.conf
```

Quite simple. 

Btw this might not the thing you're mention. Well this Dockerfile is not complete though. I'm trying to write a better `Dockerfile` and make use of `build-args` in `docker build`.

**Comment by aperture147** [comment]
I also add some build arguments, please take a glance on it

**Comment by cen1** [comment]
With folder volume mounts for each system directory I think it would be acceptable to pull this in. There is still no good way to get all the config and auxiliary files in there other than cloning and `make install` unfortunately. Maybe we could add a special cmake target to "install" those without needing to compile, if that is even possible. As a final step, storage and mysql variants could be published to hub.

**Comment by aperture147** [comment]
We could build a full function PvPGN server, but it’s kinda redundant. I think we should make some specific build to keep the server small and be optimized, like: pvpgn:mysql, pvpgn:odbc or even more specific: pvpgn:bnetd-mysql-debian or pvpgn:bnetd-mysql-ubuntu.
Btw the config file can be generated using entrypoint file. I’m trying to make it.

**Comment by RElesgoe** [comment]
@aperture147 Can you rebase this against the latest commit in the develop branch

**Comment by aperture147** [comment]
@RElesgoe hi, since this pull request, I've done some major changes:

- Try to use -march_native option (which involved changing CMakeList.txt) but after a short time using it, I've quit adding march_native so CMakeList.txt should remaining unchanged (like the upstream one)
- My Dockerfile is no longer using my github repo to build image, it now directly clone from PVPGN github repo.
- Added some instructions to README.md

So basically things I've change is README.md and Dockerfile.

**Comment by RElesgoe** [comment]
This is still a pull request for the `master` branch, can you remake this for the `develop` branch.

Additionally, it'd be nice to use the `ubuntu:latest` image instead of `buildpack-deps:bionic` and to partly match the build instructions at https://github.com/pvpgn/pvpgn-server/tree/develop#building

**Comment by aperture147** [comment]
> This is still a pull request for the `master` branch, can you remake this for the `develop` branch.
> 
> Additionally, it'd be nice to use the `ubuntu:latest` image instead of `buildpack-deps:bionic` and to partly match the build instructions at https://github.com/pvpgn/pvpgn-server/tree/develop#building

Is there any way to build a static linked PVPGN binary? If there is, then I can make a multistage Dockerfile which separate build  and run phase, reduce size of the container.

**Comment by RElesgoe** [comment]
There currently isn't an option to build static binaries.

If we want to use Ubuntu as the base image, I don't think it'll be feasible to try to compile static binaries.
In this case, it would be acceptable to use `alpine:latest` because we can conveniently install static versions of libraries with its package manager. It should then be a matter of adding `-static` to `CMAKE_CXX_FLAGS`.

**Comment by aperture147** [comment]
@RElesgoe I've just updated the dockerfile. Now the image is drastically reduced in size. Depends on user's demands (which is configurable via build args), this Dockerfile only install required libraries then remove the dev packages after the building process finished. Now the image size is reduced from 700M (using `buildpack-dev:bionic`), 400M (using `buildpack-dev:bionic-curl` or `ubuntu:bionic` with `build-essential` package), to 100M. Pretty optimized and minimized

**Comment by RElesgoe** [comment]
There are still some issues

- Can we use `ubuntu:latest` instead of `ubuntu:bionic`?
- You made a typo (`${with_pgsql}` should be `${with_odbc}`) https://github.com/aperture147/pvpgn-server/blob/e5893ec7fddbf31e3c2fc6a3edbe7d5499d51200/Dockerfile#L29
- The CMake call on [line 33](https://github.com/aperture147/pvpgn-server/blob/e5893ec7fddbf31e3c2fc6a3edbe7d5499d51200/Dockerfile#L33) still uses the undocumented command line options `-H` and `-B`. The build instructions in the [README file in the develop branch](https://github.com/pvpgn/pvpgn-server/tree/develop#generate-a-makefile-or-visual-studio-solution) works around that for people with an older version of CMake. If we use the `ubuntu:latest` image, we'll get a CMake > 3.13 which supports `-S` and `-B`.
- The README file shouldn't contain a TODO list and a known issues list, those should be in separate issues.
- The README file should have an example for creating and running separate containers for `bnetd`, `d2cs`, and `d2dbs` (would Docker Compose be useful here?).
- I'm not sure why the example in the README file mounts a single file, the `/usr/local/etc/pvpgn/ and `/usr/local/var/pvpgn/` directories should be mounted.

After addressing the above, I'd be comfortable with pulling this in.

**Comment by aperture147** [comment]
When I switched to Ubuntu 20.04/focal, MySQL 8.0 is the default version now. It seems like the current code is not compatible with `libmysqlclient-dev-8.0.22-0ubuntu0.20.04.3` and `libmysqlclient21`.
image
Btw I've already change my Dockerfile and fix all of mentioned issues above. Please check my [develop branch](https://github.com/aperture147/pvpgn-server/tree/develop)

**Comment by aperture147** [comment]
@RElesgoe Any update?

**Comment by cen1** [comment]
We should probably use `latest` for dev and ci to discover bugs like this early but for "production" uses, latest is a bad choice. I think using lts for the sake of this pull is good enough for now.

**Comment by RElesgoe** [comment]
> @RElesgoe Any update?

MySQL 8.0 is currently not supported in the `master` branch, a fix is in the `develop` branch which will be merged soon.

Also found another typo: https://github.com/aperture147/pvpgn-server/blob/70a927408d52e64e477603dab3dd77c9c96070f0/Dockerfile#L66

```
WORKDIR /usr/local/etc/pvpg
``` 

Should be:

```
WORKDIR /usr/local/sbin
```

The README file still doesn't show a complete example for creating and running 3 containers for `bnetd`, `d2cs`, and `d2dbs`. This should also be reflected in `docker-compose.yml`, such that there are 3 (and possibly 4 for MySQL) services.

Ideally, a server owner should be able to copy and paste these commands with minimal modification to run a complete PvPGN server.



> We should probably use `latest` for dev and ci to discover bugs like this early but for "production" uses, latest is a bad choice. I think using lts for the sake of this pull is good enough for now (also not to block it due to the mysql problem).

`ubuntu:latest` refers to the latest LTS release. A server owner should manually change the Dockerfile himself to specify a specific Ubuntu release to use.

**Comment by aperture147** [comment]
> 
> 
> We should probably use `latest` for dev and ci to discover bugs like this early but for "production" uses, latest is a bad choice. I think using lts for the sake of this pull is good enough for now (also not to block it due to the mysql problem).

In Ubuntu's hub, they set the `latest` to LTS, the newer version are `rolling` and `devel`. But using `latest` tag in Dockerfile is considered as a bad practice, because not every image hub in docker is label LTS version `latest`. So i think keeping the tag `focal` is good enough.

**Comment by aperture147** [comment]
I've created a new pull request #438, which is more related to this. Please check this out
